src/main/java/io/fabric8/maven/docker/CopyMojo.java

```diff
@@ -0,0 +1,271 @@
+ * with <code>docker:start</code> goal, then only the containers started by that goal are examined. Otherwise containers
+ * matching images configured in the project are searched and the copying is performed from the found containers
+ * only.</p>
+ */
+@Mojo(name = "copy", defaultPhase = LifecyclePhase.POST_INTEGRATION_TEST)
+public class CopyMojo extends AbstractDockerMojo {
+
+    private static final String COPY_NAME_PATTERN_CONFIG = "copyNamePattern";
+    private static final String TEMP_ARCHIVE_FILE_PREFIX = "docker-copy-";
+    private static final String TEMP_ARCHIVE_FILE_SUFFIX = ".tar";
+
+    /**
+     * Whether to create containers or to copy from existing containers.
+     */
+    @Parameter(property = "docker.createContainers", defaultValue = "false")
+    private boolean createContainers;
+
+    /**
+     * Naming pattern for how to name containers when created.
+     */
+    @Parameter(property = "docker.containerNamePattern")
+    private String containerNamePattern = ContainerNamingUtil.DEFAULT_CONTAINER_NAME_PATTERN;
+
+    /**
+     * Whether to copy from all containers matching configured images or only from the newest ones.
+     */
+    @Parameter(property = "docker.copyAll", defaultValue = "false")
+    private boolean copyAll;
+
+    @Override
+    protected void executeInternal(ServiceHub hub) throws IOException, MojoExecutionException {
+        DockerAccess dockerAccess = hub.getDockerAccess();
+        RunService runService = hub.getRunService();
+        ArchiveService archiveService = hub.getArchiveService();
+        GavLabel gavLabel = getGavLabel();
+        if (createContainers) {
+            log.debug("Copy mojo is invoked standalone, copying from new temporary containers");
+            copyFromTemporaryContainers(dockerAccess, runService, archiveService, gavLabel);
+        } else if (invokedTogetherWithDockerStart()) {
+            log.debug("Copy mojo is invoked together with start mojo, copying from containers created by start mojo");
+            copyFromStartedContainers(dockerAccess, runService, archiveService, gavLabel);
+        } else {
+            log.debug("Copy mojo is invoked standalone, copying from existing containers");
+            QueryService queryService = hub.getQueryService();
+            copyFromExistingContainers(dockerAccess, archiveService, queryService);
+        }
+    }
+
+    private void copyFromTemporaryContainers(final DockerAccess dockerAccess, final RunService runService,
+            final ArchiveService archiveService, final GavLabel gavLabel) throws IOException, MojoExecutionException {
+        List<ImageConfiguration> imageConfigurations = getResolvedImages();
+        for (ImageConfiguration imageConfiguration : imageConfigurations) {
+            CopyConfiguration copyConfiguration = imageConfiguration.getCopyConfiguration();
+            if (isEmpty(copyConfiguration)) {
+                continue;
```

Comment:
I think we should log here too so that user can find out that he/she is not providing CopyConfiguration

Reply:
Added debug log in 1b775a7. It's assumed that it's OK if copy configuration is missing, because user may have multiple images configured but wants to copy just from one of them. Please, let me know if you disagree and want to have info or warn log for this case.

Reply:
Thanks a lot. Looks good. I noticed this when I was trying out your PR and didn't configure any copyConfiguration. I was assuming that you implemented it like this:
```
mvn docker:copy -Ddocker.copy.containerPath=/src -Ddocker.copy.hostDir=/target
```

Reply:
This is still possible - refer to integration test in this example:

```bash
mvn -q clean install -DskipTests \
&& mvn -q -f it/builder/builder-image/pom.xml package \
&& mvn -f it/builder/builder-image/pom.xml docker:copy \
     -Ddocker.imagePropertyConfiguration=override \
     -Ddocker.createContainers  \
     -Ddocker.copy.entries.1.containerPath=//dmp-it-builder-app.tar.gz \
     -Ddocker.copy.entries.1.hostDirectory=target \
&& tar -tvf it/builder/builder-image/target/dmp-it-builder-app.tar.gz
```

Expected output looks like:

```text
[INFO] Scanning for projects...
[INFO]
[INFO] ---------< io.fabric8.dmp.itests:dmp-it-builder-builder-image >---------
[INFO] Building dmp-it-builder-builder-image 0.34-SNAPSHOT
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- docker-maven-plugin:0.34-SNAPSHOT:copy (default-cli) @ dmp-it-builder-builder-image ---
[INFO] DOCKER> Global docker.imagePropertyConfiguration=override property activates property configuration for image
[INFO] DOCKER> Copying /dmp-it-builder-app.tar.gz from 050943faf3b8 container into D:\Users\Marat\Documents\work\java\docker-maven-plugin\it\builder\builder-image\target host directory
[INFO] Expanding: C:\Users\Marat\AppData\Local\Temp\docker-copy-7878239238909315602.tar into D:\Users\Marat\Documents\work\java\docker-maven-plugin\it\builder\builder-image\target
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.996 s
[INFO] Finished at: 2021-03-23T20:42:59+03:00
[INFO] ------------------------------------------------------------------------
-rwxr-xr-x root/root     88976 1970-01-01 03:00 hello
```

Reply:
Nice, Thanks a lot for your comment. I think it would be nice if we could add all these configuration modes as documentation about `docker:copy` goal in https://github.com/fabric8io/docker-maven-plugin/tree/master/src/main/asciidoc/inc

Reply:
Please note that you still need a POM with configuration of d-m-p having at least one image, because d-m-p external configuration feature can only extend / override that image configuration, but cannot create a new image configuration.